lib/fs/fakefs_test.go

```diff
@@ -172,3 +178,722 @@ func TestFakeFSRead(t *testing.T) {
+		t.Fatal(err)
+	}
+	if fd != nil {
+		fd.Close()
+	}
+
+	fd, err = fs.OpenFile("FOOBAR", os.O_RDWR, 0664)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if fd != nil {
+		fd.Close()
+	}
+}
+
+func testFakeFSRemoveAll(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	filenames := []string{"bar", "baz", "qux"}
+	for _, filename := range filenames {
+		if _, err := fs.Create("/foo/" + filename); err != nil {
+			t.Fatalf("Could not create %s: %s", filename, err)
+		}
+	}
+
+	if err := fs.RemoveAll("/foo"); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.Stat("/foo"); err == nil {
+		t.Errorf("this should not exist anymore")
+	}
+
+	if err := fs.RemoveAll("/foo/bar"); err != nil {
+		t.Errorf("real systems don't return error here")
+	}
+}
+
+func testFakeFSRemoveAllInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	filenames := []string{"bar", "baz", "qux"}
+	for _, filename := range filenames {
+		fd, err := fs.Create("/FOO/" + filename)
+		if err != nil {
+			t.Fatalf("Could not create %s: %s", filename, err)
+		}
+		fd.Close()
+	}
+
+	if err := fs.RemoveAll("/fOo"); err != nil {
+		t.Errorf("Could not remove dir: %s", err)
+	}
+
+	if _, err := fs.Stat("/foo"); err == nil {
+		t.Errorf("this should not exist anymore")
+	}
+
+	if err := fs.RemoveAll("/foO/bAr"); err != nil {
+		t.Errorf("real systems don't return error here")
+	}
+}
+
+func testFakeFSRemove(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.Create("/Foo/Bar"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Remove("/Foo"); err == nil {
+		t.Errorf("not empty, should give error")
+	}
+
+	if err := fs.Remove("/Foo/Bar"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Remove("/Foo"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testFakeFSRemoveInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/Foo/Bar")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fd.Close()
+
+	if err := fs.Remove("/FOO"); err == nil || err == os.ErrNotExist {
+		t.Errorf("not empty, should give error")
+	}
+
+	if err := fs.Remove("/Foo/BaR"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Remove("/FoO"); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testFakeFSSameFile(t *testing.T, fs Filesystem) {
+	if runtime.GOOS == "windows" {
+		// windows time in not precise enough
+		t.SkipNow()
+	}
+
+	if err := fs.Mkdir("/Foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	filenames := []string{"Bar", "Baz", "/Foo/Bar"}
+	for _, filename := range filenames {
+		if _, err := fs.Create(filename); err != nil {
+			t.Fatalf("Could not create %s: %s", filename, err)
+		}
+	}
+
+	testCases := []struct {
+		f1   string
+		f2   string
+		want bool
+	}{
+		{f1: "Bar", f2: "Baz", want: false},
+		{f1: "Bar", f2: "/Foo/Bar", want: false},
+		{"Bar", "Bar", true},
```

Comment:
inconsistent struct init, either annotate member names, or don't

Reply:
got it